from pathlib import Path
import pandas as pd
from datetime import datetime

# Pipeline metadata
pipeline_info = {
    "created_date": "2025-05-26 17:45:47",
    "created_by": "hingelman"
}

# Print pipeline info at startup
print(f"Pipeline created by {pipeline_info['created_by']} on {pipeline_info['created_date']}")

configfile: "config/config.yaml"

# Create required directories
required_dirs = [
    "raw_data/sra",
    "raw_data/fastq",
    "trimmed_data",
    "fastqc_results",
    "alignments",
    "counts",
    "logs"
]

for path in required_dirs:
    Path(path).mkdir(parents=True, exist_ok=True)

# Read sample information
samples_df = pd.read_csv(config["samples"])
SAMPLES = samples_df['Run'].tolist()

# Print debug info
print("Samples found:", SAMPLES)

# Final output files that we want to generate
rule all:
    input:
        "counts/gene_counts.txt",
        "fastqc_results/multiqc_report.html",
        expand("alignments/{sample}.sorted.bam.bai", sample=SAMPLES)

# Download SRA files
rule prefetch_sra:
    output:
        directory("raw_data/sra/{sample}")
    log:
        "logs/prefetch/{sample}.log"
    conda:
        "envs/rnaseq.yaml"
    shell:
        "prefetch {wildcards.sample} -O {output} > {log} 2>&1"

# Extract FASTQ from SRA
rule fasterq_dump:
    input:
        "raw_data/sra/{sample}"
    output:
        r1=temp("raw_data/fastq/{sample}_1.fastq"),
        r2=temp("raw_data/fastq/{sample}_2.fastq")
    log:
        "logs/fasterq_dump/{sample}.log"
    conda:
        "envs/rnaseq.yaml"
    threads: 4
    shell:
        "fasterq-dump {input}/{wildcards.sample}.sra "
        "--threads {threads} "
        "--outdir $(dirname {output.r1}) > {log} 2>&1"

# Compress FASTQ files
rule compress_fastq:
    input:
        "raw_data/fastq/{sample}_{read}.fastq"
    output:
        "raw_data/{sample}_{read}.fastq.gz"
    log:
        "logs/pigz/{sample}_{read}.log"
    conda:
        "envs/rnaseq.yaml"
    threads: 4
    shell:
        "pigz -p {threads} -c {input} > {output} 2> {log}"

# Run FastQC
rule fastqc:
    input:
        "raw_data/{sample}_{read}.fastq.gz"
    output:
        html="fastqc_results/{sample}_{read}_fastqc.html",
        zip="fastqc_results/{sample}_{read}_fastqc.zip"
    log:
        "logs/fastqc/{sample}_{read}.log"
    conda:
        "envs/rnaseq.yaml"
    threads: config["threads"]["fastqc"]
    shell:
        "fastqc {input} -o fastqc_results -t {threads} > {log} 2>&1"

# Run MultiQC
rule multiqc:
    input:
        expand("fastqc_results/{sample}_{read}_fastqc.zip",
               sample=SAMPLES,
               read=["1", "2"])
    output:
        "fastqc_results/multiqc_report.html"
    log:
        "logs/multiqc/multiqc.log"
    conda:
        "envs/rnaseq.yaml"
    shell:
        "multiqc fastqc_results -o fastqc_results > {log} 2>&1"

# Trim reads with fastp
rule fastp:
    input:
        r1="raw_data/{sample}_1.fastq.gz",
        r2="raw_data/{sample}_2.fastq.gz"
    output:
        r1="trimmed_data/{sample}_1.trimmed.fastq.gz",
        r2="trimmed_data/{sample}_2.trimmed.fastq.gz",
        html="trimmed_data/{sample}.html",
        json="trimmed_data/{sample}.json"
    log:
        "logs/fastp/{sample}.log"
    conda:
        "envs/rnaseq.yaml"
    threads: config["threads"]["fastp"]
    params:
        extra=config["fastp"]["extra"]
    shell:
        "fastp -i {input.r1} -I {input.r2} "
        "-o {output.r1} -O {output.r2} "
        "--thread {threads} "
        "{params.extra} "
        "--html {output.html} "
        "--json {output.json} "
        "> {log} 2>&1"

# Build HISAT2 index
rule hisat2_index:
    input:
        fasta=lambda wildcards: next(Path(config["reference_dir"]).glob("*.fna"))
    output:
        multiext("reference/hisat2_index/genome",
                ".1.ht2", ".2.ht2", ".3.ht2", ".4.ht2",
                ".5.ht2", ".6.ht2", ".7.ht2", ".8.ht2")
    log:
        "logs/hisat2_index/build.log"
    conda:
        "envs/rnaseq.yaml"
    threads: config["threads"]["hisat2_build"]
    shell:
        "mkdir -p $(dirname {output[0]}) && "
        "hisat2-build {input.fasta} reference/hisat2_index/genome "
        "-p {threads} > {log} 2>&1"

# Align reads with HISAT2
rule hisat2_align:
    input:
        index=multiext("reference/hisat2_index/genome",
                      ".1.ht2", ".2.ht2", ".3.ht2", ".4.ht2",
                      ".5.ht2", ".6.ht2", ".7.ht2", ".8.ht2"),
        r1="trimmed_data/{sample}_1.trimmed.fastq.gz",
        r2="trimmed_data/{sample}_2.trimmed.fastq.gz"
    output:
        temp("alignments/{sample}.unsorted.bam")
    log:
        "logs/hisat2/{sample}.log"
    conda:
        "envs/rnaseq.yaml"
    threads: config["threads"]["hisat2_align"]
    params:
        extra=config["hisat2"]["extra"]
    shell:
        "hisat2 -p {threads} "
        "-x reference/hisat2_index/genome "
        "-1 {input.r1} -2 {input.r2} "
        "{params.extra} "
        "2> {log} | "
        "samtools view -bS - > {output}"

# Sort BAM files
rule samtools_sort:
    input:
        "alignments/{sample}.unsorted.bam"
    output:
        "alignments/{sample}.sorted.bam"
    log:
        "logs/samtools_sort/{sample}.log"
    conda:
        "envs/rnaseq.yaml"
    threads: config["threads"]["samtools_sort"]
    shell:
        "samtools sort -@ {threads} "
        "-o {output} {input} > {log} 2>&1"

# Index BAM files
rule samtools_index:
    input:
        "alignments/{sample}.sorted.bam"
    output:
        "alignments/{sample}.sorted.bam.bai"
    log:
        "logs/samtools_index/{sample}.log"
    conda:
        "envs/rnaseq.yaml"
    shell:
        "samtools index {input} > {log} 2>&1"

# Count reads with featureCounts
rule featurecounts:
    input:
        bams=expand("alignments/{sample}.sorted.bam", sample=SAMPLES),
        gtf=lambda wildcards: next(Path(config["reference_dir"]).glob("*.gtf"))
    output:
        counts="counts/gene_counts.txt",
        summary="counts/gene_counts.txt.summary"
    log:
        "logs/featurecounts/counts.log"
    conda:
        "envs/rnaseq.yaml"
    threads: config["threads"]["featurecounts"]
    params:
        feature_type=config["featurecounts"]["feature_type"],
        attribute_type=config["featurecounts"]["attribute_type"],
        extra=config["featurecounts"]["extra"]
    shell:
        "featureCounts "
        "-T {threads} "
        "-t {params.feature_type} "
        "-g {params.attribute_type} "
        "{params.extra} "
        "-a {input.gtf} "
        "-o {output.counts} "
        "{input.bams} > {log} 2>&1"